### ISSUE TITLE
Inline commenting

### ISSUE THREAD

**Issue by abbeyjackson** [BookStackApp/BookStack#1265]
This is a suggestion added to issues as per Dan’s instructions via Twitter. 

Inline commenting (selecting a chubk of text to comment on or having a comment associated with a specific line number) is a feature that is missing from Bookstack which I feel would really add value to the product.

**Comment by JohnnyWild89** [comment]
Looking forward to get this feature! Bookshelf is perfect replacement for any products which use as WIKI. Inline commentaries would greatly expands work possibilities.

**Comment by cze-karel** [comment]
This would be a really great feature that would improve the possibility of people collaborating on the documents. Is it possible this feature will make it into the roadmap?

**Comment by ssddanbrown** [comment]
Further request and ideas proposed within #3063

**Comment by TiemeWoldman** [comment]
Any plans on adding annotations (Inline comments) to BookStack? This would be a great addition. 

I tried [hypothes.is](https://hypothes.is/) for adding annotations to BookStack pages, but the BookStack text select pop-up interferes with hypothes.

**Comment by EDMJohnny** [comment]
As far as I can tell BookStack is pretty close to perfect for a writing / collaboration platform. Super simple to use, very intuitive.  Inline comments would make it perfect!  Is this something you are working on or have an ETA for?
Thanks

**Comment by ssddanbrown** [comment]
@EDMJohnny

>  Is this something you are working on or have an ETA for?

Not at this time.

**Comment by EDMJohnny** [comment]
Thanks for responding.  Is this something you'd be willing to put a bounty on / get paid for completing?  Do you have a ballpark for cost of adding this feature?

**Comment by ssddanbrown** [comment]
@EDMJohnny I'm not really a fan of public bounties, can just cause issues. Not really sure on ballpark costs either as this would need fleshing out into a full proposal since this request could be imagined/desired in different forms.

I've quoted features before (although never accepted) but only for features with clear scope and platform fit. Something makes me uneasy about rushing through a proposal for the sake of money, feel that risks providing an unbalanced larger voice to those with money, not sure the right way to balance that. If I was to quote it'd be somewhere upwards of £1k, even for a minimal implementation.

**Comment by danhansen** [comment]
Why not integrate something like https://github.com/recogito/recogito-js ? It looks like a significant portion of the functionality is already implemented.

**Comment by BloodyIron** [comment]
@ssddanbrown can we please have this added to a roadmap in some regard? As you may remember seeing me elsewhere, likely to switch to Bookstack, and some of our staff use this kind of feature in Confuence (which we are moving away from).

Is there a hurdle on your end for this?

**Comment by ssddanbrown** [comment]
> Is there a hurdle on your end for this?

Uh, I guess it's some level of formula of:

```
PRESSURE = ("user actual benefit" + "user demand" + "my own enthusiasm level for the feature")
DRAG = ("implementation cost" + "my predicated cost of maintenance, inc. support")
HURDLE = ((PRESSURE - DRAG) * "my current emotions" * RANDOM_CHANCE())
```

That's probably the best I can explain it, same for any feature tbh. 
Roadmap is reserved for high-level elements. Could maybe have some kind of "Commenting and Communications" item. Not sure though, might want to address other things before.

Realistically, the next step for this specific request would be for me to put forward an implementation proposal, with a scope that I'd be happy to maintain, for community feedback. No reason to rush though IMO.

**Comment by ironbelly** [comment]
Agreed with EDMJohnny above that inline comments to facilitate more specific conversation amongst collaborators is one of the biggest items of value that could be added.

Also inline with EDMJohnny about the bounty, would be happy to chip in

**Comment by EDMJohnny** [comment]
@ssddanbrown  Hi - just checking in again.  I see you've done a lot of updates on Bookstack since I last checked it.  All much appreciated.  Any chance of bumping this feature request for inline comments on to your To Do?

**Comment by ipimpat** [comment]
This would be a killer feature !

**Comment by rosequaartz** [comment]
This feature would be great! :)

**Comment by shashankvivek** [comment]
Yeah, with this feature, bookstack can knock the ball out of the park !

**Comment by stevenvegt** [comment]
I stumbled on this issue and wanted to chip in. We use bookstack for our open source community. People are working together on projects and keep reverting to google docs. Inline comments are the missing key in bookstack and holding them back to make the switch. I hope this will add some weight to the` "user actual benefit" + "user demand"` parts of the equation. Either way, thanks for this nice piece of software!

**Comment by dtufood-kihen** [comment]
Just dropping by to say I've been keeping an eye on the chat about inline comments. I get that developers have to juggle a bunch of stuff and can't just jump on every new idea, but I can't help feeling pretty stoked about this particular feature.

Please make it happen ! :-)

**Comment by khudari-starpath** [comment]
+1 would switch to bookstack if it had this feature!

**Comment by agapoar** [comment]
Hi. I would also love to see this feature added. By way of suggestion, I expected to see the ability to add an inline comment when I double clicked on a word in the text in read mode. I did see a popup menu, which was great. But obviously it didn't contain the ability to annotate. This screenshot shows the menu.

![image](https://github.com/BookStackApp/BookStack/assets/166620720/837f59f5-ed58-46dc-9ba5-6a2d50264a21)

**Comment by ipimpat** [comment]
Hi @ssddanbrown,

First, I want to express my gratitude for the incredible work.

I was wondering if you might consider adding the annotation feature to the roadmap. The demand for this feature seems quite high among users.

Currently, the absence of this feature is the main reason my team continues to rely on Word files. They are hesitant to transition to tools that lack annotation capabilities.

Thank you so much for considering this request, and for all the amazing work you continue to do.

**Comment by bendem** [comment]
> The demand for this feature seems quite high among users.

If you look at the most requested feature, it has >100 👍🏻. This one has 25.

Reminder that adding a 👍🏻 on the top post is the way to vote for an issue, pressuring/guilting the developer is not the way to get the feature you want added in.

![image](https://github.com/BookStackApp/BookStack/assets/2681677/4b5d3017-93af-4826-ac65-df1d0e5c4003)

**Comment by grv688** [comment]
Also very interested in this.  We're migrating from documents that allowed inline commenting during revision cycles, and it was very useful.

**Comment by pato-lama** [comment]
Tossing a comment in here to support this request - would love having this functionality!

**Comment by ahnchive** [comment]
Also very interested in this!

**Comment by Yarrax** [comment]
Also very interested in this!

**Comment by Yarrax** [comment]
Hmmm... I'm sorry, but this feature is looks like most wanted:

![Image](https://github.com/user-attachments/assets/ce26bccb-aae9-4e3f-aef7-3672b3e39931)

Of course, it haven't >100 thumbs up, but... :)

**Comment by sjimellor-jr** [comment]
Is Hypothes.is (https://web.hypothes.is/) an option? It can already be used to privately annotate Bookstack pages with an account and browser extension. Perhaps with a self-hosted "h" server (https://github.com/hypothesis/h) and using the JS client (https://github.com/hypothesis/client) configured to use the self-hosted server then this could be more native. I've not tried to host my own "h" server.

**Comment by Yarrax** [comment]
> Is Hypothes.is (https://web.hypothes.is/) an option?

Hmm... may be, thanks i'll try it. but anyway it will be a one more service for manual install. It will be great if bookstack will have inline commenting onboard

**Comment by ssddanbrown** [comment]
I've now added an initial implementation for in-content comments via PR #5584.
This is due to be part of the next feature release.
Therefore I'm going to close this off. Further related requests, or issues upon the implementation, can be raised in future requests if needed (once there's been opportunity to test after the next feature release).

This implementation is (relatively) minimal so may have some limitations in the view of some, but otherwise it should achieve the general functional goals for most. This implementation includes:

- Ability to start a content-connected comment thread from the content popover when viewing a page.
- The comment references a specific block of content, along with an optional text range within that to highlight a range of text.
- Comment threads will show as highlighted ranges when viewing a page (as long as the referenced sections exist).
  - You can click the comment icon within a highlight range to open up the comment thread inline.
  - There's a button to jump to the full thread for interaction (reply, edit, delete etc...).
- Comments will also show inline within the new (in-alpha) WYSIWYG editor when the comments sidebar is opened in the editor view.
- To support this functionality, I've also added the ability to archive (and unarchive) comments, so that these kind of comments can be treated as a kind of todo-list, and archived when addressed. 
  - Active/Archived threads are listed under separate tabs.
  - Inline comments will only be shown for the active comments tab.
- Comments are marked if they have a reference.
  - This can be clicked to jump the the referenced section.
  - This will update to show as outdated if the reference section has changed.
    - This is based upon a hash of the text for that block section.
  - This will turn orange if the section no longer exists at all.
  - This will be gray in scenarios where we can't provide an interactive link or check the reference (in MD/old-wyswiyg editors)

**Limitations**

- Only the top-level comment in a thread can include a reference (and also be archived/unarchived).
- Only the new WYSIWYG editor receives inline support.
- The inline comments are not interactive (no reply or other controls) but they do have the ability to jump to the full comment.
- There's no display options/controls for these comments. Could be something that's needed but I thought we'd get the feature out there first to understand what level of control is actually needed (if at all).
- The outdated/change detection is currently done browser-side, and is done block level (not just the text range if set).
  - I've tried to do this in a way we should be able to replicate on the PHP side if needed in the future.

### Screenshots

<details><summary>Inline Display</summary>
<p>

![Image](https://github.com/user-attachments/assets/f1bb78dc-f636-4a10-a081-9b346ce5c30f)

</p>
</details> 

<details><summary>Inline Display - Opened</summary>
<p>

![Image](https://github.com/user-attachments/assets/e6778242-588c-4a8b-ad76-3dd9a95fb4bb)

</p>
</details> 

<details><summary>Page View Comments List</summary>
<p>

![Image](https://github.com/user-attachments/assets/7b2c0945-53a8-44a1-b611-d3e359e0a9d4)

</p>
</details> 

<details><summary>In Editor</summary>
<p>

![Image](https://github.com/user-attachments/assets/528d4697-d249-48b9-8613-242a765258a2)

</p>
</details>

**Comment by pabalven** [comment]
Thank you so much @ssddanbrown for releasing this. I just saw that in the demo site. Great implementation, and so useful!

Thanks, thanks, thanks!!!
